File: EnvironmentMap/EnvironmentMap/Renderer.m
//
//  Renderer.m
//  4.8.EnvironmentMap
//
//  Created by Jacob Su on 3/24/21.
//

@import MetalKit;
#import <Foundation/Foundation.h>
#import "Renderer.h"
#import <common/common.h>
#import "ShaderType.h"
#include <simd/simd.h>

static const float PI = 3.1415926;

@implementation Renderer
{
    Camera *_camera;
    id<MTLDevice> _device;
    Uniforms _uniform;
    id<MTLTexture> _skyBoxTexture;
    id<MTLTexture> _depthTexture;
    id<MTLRenderPipelineState> _boxRenderPipelineState;
    id<MTLRenderPipelineState> _skyRenderPipelineState;
    id<MTLDepthStencilState> _lessEqualDepthState;
    id<MTLDepthStencilState> _lessDepthState;
    vector_uint2 _viewportSize;
    id<MTLCommandQueue> _commandQueue;
    MTKMesh *_cubeMesh;
    id<MTLBuffer> _skyBoxBuffer;
}

- (nonnull instancetype)initWithMetalKitView:(nonnull MTKView*)mtkView {
    self = [super init];
    if (self) {
        _device = mtkView.device;
        mtkView.delegate = self;
        mtkView.depthStencilPixelFormat = MTLPixelFormatDepth32Float;
        
        _camera = [[Camera alloc] initWithPosition:(vector_float3) {0.0, 0.0, 5.0}
                                        withTarget:(vector_float3) {0.0, 0.0, 0.0}
                                            withUp:(vector_float3) {0.0, 1.0, 0.0}];
        
        MTLVertexDescriptor *mtlVertexDescriptor = [MTLVertexDescriptor new];
        mtlVertexDescriptor.attributes[VertexAttributeIndexPosition].format = MTLVertexFormatFloat3;
        mtlVertexDescriptor.attributes[VertexAttributeIndexPosition].offset = 0;
        mtlVertexDescriptor.attributes[VertexAttributeIndexPosition].bufferIndex = VertexInputIndexPosition;
        
        mtlVertexDescriptor.attributes[VertexAttributeIndexNormal].format = MTLVertexFormatFloat3;
        mtlVertexDescriptor.attributes[VertexAttributeIndexNormal].offset = 16;
        mtlVertexDescriptor.attributes[VertexAttributeIndexNormal].bufferIndex = VertexInputIndexPosition;
        
        mtlVertexDescriptor.layouts[VertexInputIndexPosition].stride = 32;
        mtlVertexDescriptor.layouts[VertexInputIndexPosition].stepRate = 1;
        mtlVertexDescriptor.layouts[VertexInputIndexPosition].stepFunction = MTLStepFunctionPerVertex;
        
        int width = mtkView.frame.size.width;
        int height = mtkView.frame.size.height;
        
        _depthTexture = [self buildDepthTextureWithWidth:width height:height];
        id<MTLLibrary> library = [_device newDefaultLibrary];
        id<MTLFunction> vertexFunc = [library newFunctionWithName:@"vertexShader"];
        id<MTLFunction> fragmentFunc = [library newFunctionWithName:@"fragmentShader"];
        
        MTLRenderPipelineDescriptor *boxPipelineDescriptor = [MTLRenderPipelineDescriptor new];
        boxPipelineDescriptor.label = @"box pipeline";
        boxPipelineDescriptor.vertexFunction = vertexFunc;
        boxPipelineDescriptor.fragmentFunction = fragmentFunc;
        boxPipelineDescriptor.vertexDescriptor = mtlVertexDescriptor;
        boxPipelineDescriptor.colorAttachments[0].pixelFormat = mtkView.colorPixelFormat;
        boxPipelineDescriptor.depthAttachmentPixelFormat = _depthTexture.pixelFormat;
        
        NSError *error;
        _boxRenderPipelineState = [_device newRenderPipelineStateWithDescriptor:boxPipelineDescriptor error:&error];
        
        NSAssert(_boxRenderPipelineState, @"new box render pipeline state error %@", error);
        
        id<MTLFunction> cubeMapVertexFunc = [library newFunctionWithName:@"cubeMapVertexShader"];
        id<MTLFunction> cubeMapFragmentFunc = [library newFunctionWithName:@"cubeMapFragmentShader"];
        
        MTLRenderPipelineDescriptor *skyPipelineDescriptor = [MTLRenderPipelineDescriptor new];
        skyPipelineDescriptor.label = @"sky pipeline";
        skyPipelineDescriptor.vertexFunction = cubeMapVertexFunc;
        skyPipelineDescriptor.fragmentFunction = cubeMapFragmentFunc;
        skyPipelineDescriptor.colorAttachments[0].pixelFormat = mtkView.colorPixelFormat;
        skyPipelineDescriptor.depthAttachmentPixelFormat = _depthTexture.pixelFormat;
        
        _skyRenderPipelineState = [_device newRenderPipelineStateWithDescriptor:skyPipelineDescriptor error:&error];
        NSAssert(_skyRenderPipelineState, @"new sky render pipeline state error %@", error);
        
        MTLDepthStencilDescriptor *lessEqualDepthDescriptor = [MTLDepthStencilDescriptor new];
        lessEqualDepthDescriptor.depthCompareFunction = MTLCompareFunctionLessEqual;
        lessEqualDepthDescriptor.depthWriteEnabled = YES;
        
        _lessEqualDepthState = [_device newDepthStencilStateWithDescriptor:lessEqualDepthDescriptor];
        
        MTLDepthStencilDescriptor *lessDepthDescriptor = [MTLDepthStencilDescriptor new];
        lessDepthDescriptor.depthCompareFunction = MTLCompareFunctionLess;
        lessDepthDescriptor.depthWriteEnabled = YES;
        
        _lessDepthState = [_device newDepthStencilStateWithDescriptor:lessDepthDescriptor];
        
        _viewportSize = (vector_uint2) { mtkView.frame.size.width, mtkView.frame.size.height };
        _commandQueue = [_device newCommandQueue];
        
        _cubeMesh = [MTKMesh newBoxWithVertexDescriptor:mtlVertexDescriptor
                                      withAttributesMap:@{
                                          [NSNumber numberWithInt:VertexAttributeIndexPosition]: MDLVertexAttributePosition,
                                          [NSNumber numberWithInt:VertexAttributeIndexNormal]: MDLVertexAttributeNormal
                                                    }
                                             withDevice:_device
                                         withDimensions:(vector_float3) {1.0, 1.0, 1.0}
                                               segments:(vector_uint3) {1, 1, 1}
                                           geometryType:MDLGeometryTypeTriangles
                                          inwardNormals:false
                                                  error:&error];
        NSAssert(_cubeMesh, @"cube mesh error: %@", error);
        
        CubeMapVertex skyboxVertices[] =
        {
            {{-1.0,  1.0, -1.0}},
            {{-1.0, -1.0, -1.0}},
            {{ 1.0, -1.0, -1.0}},
            {{ 1.0, -1.0, -1.0}},
            {{ 1.0,  1.0, -1.0}},
            {{-1.0,  1.0, -1.0}},
            
            {{-1.0, -1.0,   1.0}},
            {{-1.0, -1.0,  -1.0}},
            {{-1.0,  1.0,  -1.0}},
            {{-1.0,  1.0,  -1.0}},
            {{-1.0,  1.0,   1.0}},
            {{-1.0, -1.0,   1.0}},
            
            {{ 1.0, -1.0, -1.0}},
            {{ 1.0, -1.0,  1.0}},
            {{ 1.0,  1.0,  1.0}},
            {{ 1.0,  1.0,  1.0}},
            {{ 1.0,  1.0, -1.0}},
            {{ 1.0, -1.0, -1.0}},

            {{-1.0, -1.0,  1.0}},
            {{-1.0,  1.0,  1.0}},
            {{ 1.0,  1.0,  1.0}},
            {{ 1.0,  1.0,  1.0}},
            {{ 1.0, -1.0,  1.0}},
            {{-1.0, -1.0,  1.0}},
            
            {{-1.0,  1.0, -1.0}},
            {{ 1.0,  1.0, -1.0}},
            {{ 1.0,  1.0,  1.0}},
            {{ 1.0,  1.0,  1.0}},
            {{-1.0,  1.0,  1.0}},
            {{-1.0,  1.0, -1.0}},

            {{-1.0, -1.0, -1.0}},
            {{-1.0, -1.0,  1.0}},
            {{ 1.0, -1.0, -1.0}},
            {{ 1.0, -1.0, -1.0}},
            {{-1.0, -1.0,  1.0}},
            {{ 1.0, -1.0,  1.0}},
        };
        
        _skyBoxBuffer = [_device newBufferWithBytes:skyboxVertices
                                             length:sizeof(skyboxVertices)
                                            options:MTLResourceStorageModeShared];
        
        _skyBoxTexture = [TextureCubeLoader
                            loadWithImageNames:@[
                                @"skybox/right.jpg",
                                @"skybox/left.jpg",
                                @"skybox/top.jpg",
                                @"skybox/bottom.jpg",
                                @"skybox/front.jpg",
                                @"skybox/back.jpg" ]
                          device:_device
                          commandQueue:_commandQueue
                          error:&error];
        _uniform.cameraPos = [_camera getCameraPosition];
        _uniform.viewMatrix = [_camera getViewMatrix];
        _uniform.modelMatrix = matrix4x4_identity();
        _uniform.projectionMatrix = matrix_perspective_left_hand(PI / 4.0, (float) width / (float) height, 0.1, 100.0);;
        _uniform.inverseModelMatrix = simd_inverse(_uniform.modelMatrix);
    }
    
    return self;
}

- (id<MTLTexture>) buildDepthTextureWithWidth:(int) width height:(int) height {
    MTLTextureDescriptor *descriptor = [MTLTextureDescriptor new];
   
    descriptor.pixelFormat = MTLPixelFormatDepth32Float;
    descriptor.width = width;
    descriptor.height = height;
    descriptor.mipmapLevelCount = 1;
    descriptor.usage = MTLTextureUsageRenderTarget;
    descriptor.storageMode = MTLStorageModePrivate;
    
    id<MTLTexture> depthTexture = [_device newTextureWithDescriptor:descriptor];
    depthTexture.label = @"Depth Texture";
    
    return depthTexture;
}

- (void) handleMouseScrollDeltaX:(float) deltaX deltaY:(float) deltaY {
    [_camera handleMouseScrollDeltaX:deltaX deltaY:deltaY];
    
    _uniform.cameraPos = [_camera getCameraPosition];
    _uniform.viewMatrix = [_camera getViewMatrix];
}

- (void)drawInMTKView:(nonnull MTKView *)view {
    view.clearDepth = 1.0;
    
    id<MTLCommandBuffer> commandBuffer = [_commandQueue commandBuffer];
    
    MTLRenderPassDescriptor *renderDescriptor = [MTLRenderPassDescriptor new];
    renderDescriptor.colorAttachments[0].texture = [[view currentDrawable] texture];
    renderDescriptor.colorAttachments[0].loadAction = MTLLoadActionClear;
    renderDescriptor.colorAttachments[0].storeAction = MTLStoreActionStore;
    renderDescriptor.colorAttachments[0].clearColor = MTLClearColorMake(0.2, 0.3, 0.3, 1.0);
    
    renderDescriptor.depthAttachment.texture = _depthTexture;
    renderDescriptor.depthAttachment.loadAction = MTLLoadActionClear;
    renderDescriptor.depthAttachment.storeAction = MTLStoreActionDontCare;
    renderDescriptor.depthAttachment.clearDepth = 1.0;
    
    id<MTLRenderCommandEncoder> boxRenderEncoder = [commandBuffer renderCommandEncoderWithDescriptor:renderDescriptor];
    boxRenderEncoder.label = @"box render";
    [boxRenderEncoder setViewport:(MTLViewport) {0.0, 0.0, _viewportSize.x, _viewportSize.y, 0.0, 1.0}];
    
    // draw box
    [boxRenderEncoder setRenderPipelineState:_boxRenderPipelineState];
    [boxRenderEncoder setDepthStencilState:_lessDepthState];

    for (int i = 0; i < _cubeMesh.vertexBuffers.count; i++) {
        [boxRenderEncoder setVertexBuffer:_cubeMesh.vertexBuffers[i].buffer offset:_cubeMesh.vertexBuffers[i].offset atIndex:VertexInputIndexPosition];
    }

    [boxRenderEncoder setVertexBytes:&_uniform length:sizeof(_uniform) atIndex:VertexInputIndexUniforms];
    [boxRenderEncoder setFragmentTexture:_skyBoxTexture atIndex:FragmentInputIndexCubeTexture];

    for (int i = 0; i < _cubeMesh.submeshes.count; i++) {
        MTKSubmesh *mesh = _cubeMesh.submeshes[i];
        [boxRenderEncoder drawIndexedPrimitives:mesh.primitiveType
                                     indexCount:mesh.indexCount
                                      indexType:mesh.indexType
                                    indexBuffer:mesh.indexBuffer.buffer
                              indexBufferOffset:mesh.indexBuffer.offset];
    }
    
    // draw sky
    [boxRenderEncoder setRenderPipelineState:_skyRenderPipelineState];
    [boxRenderEncoder setDepthStencilState:_lessEqualDepthState];

    [boxRenderEncoder setVertexBuffer:_skyBoxBuffer offset:0 atIndex:VertexInputIndexPosition];
    [boxRenderEncoder drawPrimitives:MTLPrimitiveTypeTriangle vertexStart:0 vertexCount:36];
    
    [boxRenderEncoder endEncoding];
    
    [commandBuffer presentDrawable:view.currentDrawable];
    [commandBuffer commit];
    [commandBuffer waitUntilCompleted];
}

- (void)mtkView:(nonnull MTKView *)view drawableSizeWillChange:(CGSize)size {
    _viewportSize.x = size.width;
    _viewportSize.y = size.height;
    
    if (_depthTexture.width != size.width ||
            _depthTexture.height != size.height)  {
        _depthTexture = [self buildDepthTextureWithWidth:size.width height:size.height];
    }

    _uniform.projectionMatrix = matrix_perspective_left_hand(PI / 4.0, (float) size.width / (float) size.height, 0.1, 100.0);
}

@end

File: CubeMaps/CubeMaps/Renderer.swift
//
//  Renderer.swift
//  4.7.CubeMaps
//
//  Created by Jacob Su on 3/23/21.
//

import Foundation
import common
import Metal
import MetalKit

let PI : Float = 3.1415926

class Renderer : NSObject {
    
    private var device: MTLDevice!
    private var camera: Camera!
    
    private var renderPipelineState: MTLRenderPipelineState!
    private var skyBoxPipelineState: MTLRenderPipelineState!
    private var lessEqualDepthState: MTLDepthStencilState!
    private var lessDepthState: MTLDepthStencilState!
    private var commandQueue: MTLCommandQueue!
    private var viewportSize: vector_uint2!
    private var cubeMesh: MTKMesh!
    private var skyBoxBuffer: MetalBuffer<CubeMapVertex>!
    private var skyBoxTexture: MTLTexture!
    private var depthTexture: MTLTexture!
    private var containerTexture: Texture!
    private var uniform: Uniforms!
    private let depthFormat: MTLPixelFormat = .depth32Float

    init(metalView: MTKView) {
        super.init()
        
        device = metalView.device!
        metalView.delegate = self
        
        camera = Camera(position: vector_float3(0.0, 0.0, 3.0),
                        withTarget: vector_float3(0.0, 0.0, 0.0),
                        withUp: vector_float3(0.0, 1.0, 0.0))
        
        let mtlVertexDescriptor = MTLVertexDescriptor()
        // positions
        mtlVertexDescriptor.attributes[Int(VertexAttributeIndexPosition.rawValue)].format = MTLVertexFormat.float3
        mtlVertexDescriptor.attributes[Int(VertexAttributeIndexPosition.rawValue)].offset = 0
        mtlVertexDescriptor.attributes[Int(VertexAttributeIndexPosition.rawValue)].bufferIndex = Int(VertexInputIndexPosition.rawValue)
        
        // texture coordinates
        mtlVertexDescriptor.attributes[Int(VertexAttributeIndexTexcoord.rawValue)].format = MTLVertexFormat.float2
        mtlVertexDescriptor.attributes[Int(VertexAttributeIndexTexcoord.rawValue)].offset = 16
        mtlVertexDescriptor.attributes[Int(VertexAttributeIndexTexcoord.rawValue)].bufferIndex = Int(VertexInputIndexPosition.rawValue)
        
        mtlVertexDescriptor.layouts[Int(VertexInputIndexPosition.rawValue)].stride = 32
        mtlVertexDescriptor.layouts[Int(VertexInputIndexPosition.rawValue)].stepRate = 1
        mtlVertexDescriptor.layouts[Int(VertexInputIndexPosition.rawValue)].stepFunction = MTLVertexStepFunction.perVertex
        
        let library = device.makeDefaultLibrary()!
        let vertexFunc = library.makeFunction(name: "vertexShader")!
        let fragmentFunc = library.makeFunction(name: "fragmentShader")!
        
        let pipelineDescriptor = MTLRenderPipelineDescriptor()
        pipelineDescriptor.label = "cube container"
        pipelineDescriptor.vertexFunction = vertexFunc
        pipelineDescriptor.fragmentFunction = fragmentFunc
        pipelineDescriptor.vertexDescriptor = mtlVertexDescriptor
        pipelineDescriptor.colorAttachments[0].pixelFormat = metalView.colorPixelFormat
        pipelineDescriptor.depthAttachmentPixelFormat = depthFormat
        
        renderPipelineState = try! device.makeRenderPipelineState(descriptor: pipelineDescriptor)
        
        let lessEqualDepthStateDescritpor = MTLDepthStencilDescriptor()
        lessEqualDepthStateDescritpor.depthCompareFunction = .lessEqual
        lessEqualDepthStateDescritpor.isDepthWriteEnabled = true
        
        lessEqualDepthState = device.makeDepthStencilState(descriptor: lessEqualDepthStateDescritpor)
        
        let lessDepthStateDescriptor = MTLDepthStencilDescriptor()
        lessDepthStateDescriptor.depthCompareFunction = .less
        lessDepthStateDescriptor.isDepthWriteEnabled = true
        
        lessDepthState = device.makeDepthStencilState(descriptor: lessDepthStateDescriptor)
        
        let cubeMapVertexFunc = library.makeFunction(name: "cubeMapVertexShader")
        let cubeMapFragmentFunc = library.makeFunction(name: "cubeMapFragmentShader")
        
        let skyBoxPipelineDescriptor = MTLRenderPipelineDescriptor()
        skyBoxPipelineDescriptor.label = "sky box"
        skyBoxPipelineDescriptor.vertexFunction = cubeMapVertexFunc
        skyBoxPipelineDescriptor.fragmentFunction = cubeMapFragmentFunc
        skyBoxPipelineDescriptor.colorAttachments[0].pixelFormat = metalView.colorPixelFormat
        skyBoxPipelineDescriptor.depthAttachmentPixelFormat = depthFormat
        
        skyBoxPipelineState = try! device.makeRenderPipelineState(descriptor: skyBoxPipelineDescriptor)
        
        commandQueue = device.makeCommandQueue()!
        viewportSize = vector_uint2(UInt32(metalView.frame.width), UInt32(metalView.frame.height))
        
        cubeMesh = try! MTKMesh.newBox(withVertexDescriptor: mtlVertexDescriptor,
                                       withAttributesMap: [
            Int(VertexAttributeIndexPosition.rawValue): MDLVertexAttributePosition,
            Int(VertexAttributeIndexTexcoord.rawValue): MDLVertexAttributeTextureCoordinate],
                                       withDevice: device)
        
        let skyboxVertices: [CubeMapVertex] = [
            CubeMapVertex(position: vector_float3(-1.0,  1.0, -1.0)),
            CubeMapVertex(position: vector_float3(-1.0, -1.0, -1.0)),
            CubeMapVertex(position: vector_float3( 1.0, -1.0, -1.0)),
            CubeMapVertex(position: vector_float3( 1.0, -1.0, -1.0)),
            CubeMapVertex(position: vector_float3( 1.0,  1.0, -1.0)),
            CubeMapVertex(position: vector_float3(-1.0,  1.0, -1.0)),
            
            CubeMapVertex(position: vector_float3(-1.0, -1.0,  1.0)),
            CubeMapVertex(position: vector_float3(-1.0, -1.0, -1.0)),
            CubeMapVertex(position: vector_float3(-1.0,  1.0, -1.0)),
            CubeMapVertex(position: vector_float3(-1.0,  1.0, -1.0)),
            CubeMapVertex(position: vector_float3(-1.0,  1.0,  1.0)),
            CubeMapVertex(position: vector_float3(-1.0, -1.0,  1.0)),

            CubeMapVertex(position: vector_float3( 1.0, -1.0, -1.0)),
            CubeMapVertex(position: vector_float3( 1.0, -1.0,  1.0)),
            CubeMapVertex(position: vector_float3( 1.0,  1.0,  1.0)),
            CubeMapVertex(position: vector_float3( 1.0,  1.0,  1.0)),
            CubeMapVertex(position: vector_float3( 1.0,  1.0, -1.0)),
            CubeMapVertex(position: vector_float3( 1.0, -1.0, -1.0)),
            
            CubeMapVertex(position: vector_float3(-1.0, -1.0,  1.0)),
            CubeMapVertex(position: vector_float3(-1.0,  1.0,  1.0)),
            CubeMapVertex(position: vector_float3( 1.0,  1.0,  1.0)),
            CubeMapVertex(position: vector_float3( 1.0,  1.0,  1.0)),
            CubeMapVertex(position: vector_float3( 1.0, -1.0,  1.0)),
            CubeMapVertex(position: vector_float3(-1.0, -1.0,  1.0)),

            CubeMapVertex(position: vector_float3(-1.0,  1.0, -1.0)),
            CubeMapVertex(position: vector_float3( 1.0,  1.0, -1.0)),
            CubeMapVertex(position: vector_float3( 1.0,  1.0,  1.0)),
            CubeMapVertex(position: vector_float3( 1.0,  1.0,  1.0)),
            CubeMapVertex(position: vector_float3(-1.0,  1.0,  1.0)),
            CubeMapVertex(position: vector_float3(-1.0,  1.0, -1.0)),

            CubeMapVertex(position: vector_float3(-1.0, -1.0, -1.0)),
            CubeMapVertex(position: vector_float3(-1.0, -1.0,  1.0)),
            CubeMapVertex(position: vector_float3( 1.0, -1.0, -1.0)),
            CubeMapVertex(position: vector_float3( 1.0, -1.0, -1.0)),
            CubeMapVertex(position: vector_float3(-1.0, -1.0,  1.0)),
            CubeMapVertex(position: vector_float3( 1.0, -1.0,  1.0)),
        ]
        
        skyBoxBuffer = MetalBuffer(device: device,
                                   array: skyboxVertices,
                                   index: UInt32(VertexInputIndexPosition.rawValue))
        
        containerTexture = try! Texture.newTextureWithName("container", scaleFactor: 1.0, device: device, index: Int(FragmentInputIndexDiffuseTexture.rawValue))
        
        skyBoxTexture = try! TextureCubeLoader.load(withImageNames: [
                                                "skybox/right.jpg",
                                                "skybox/left.jpg",
                                                "skybox/top.jpg",
                                                "skybox/bottom.jpg",
                                                "skybox/front.jpg",
                                                "skybox/back.jpg"],
                                               device: device, commandQueue: commandQueue)
        let width = metalView.frame.width
        let height = metalView.frame.height
        
        depthTexture = buildDepthTexture(Int(width), Int(height))
        
        uniform = Uniforms(
            modelMatrix: matrix4x4_identity(),
            viewMatrix: camera.getViewMatrix(),
            projectionMatrix: matrix_perspective_left_hand(PI / 4.0, Float(width) / Float(height), 0.1, 100.0))
    }
    
    func handleCameraEvent(deltaX: Float, deltaY: Float) -> Void {
        camera.handleMouseScrollDeltaX(deltaX, deltaY: deltaY)
        
        uniform.viewMatrix = camera.getViewMatrix()
    }
    
    fileprivate func buildDepthTexture(_ width: Int, _ height: Int) -> MTLTexture {
        let descriptor = MTLTextureDescriptor()
        descriptor.pixelFormat = depthFormat
        descriptor.width = width
        descriptor.height = height
        descriptor.mipmapLevelCount = 1
        descriptor.usage = .renderTarget
        descriptor.storageMode = .private
        
        let depthTexture = device.makeTexture(descriptor: descriptor)!
        depthTexture.label = "Depth Texture"
        return depthTexture
    }
}

extension Renderer: MTKViewDelegate {
    func mtkView(_ view: MTKView, drawableSizeWillChange size: CGSize) {
        viewportSize.x = UInt32(size.width)
        viewportSize.y = UInt32(size.height)
        
        if (depthTexture.width != Int(size.width) ||
                depthTexture.height != Int(size.height))  {
            depthTexture = buildDepthTexture(Int(size.width), Int(size.height))
        }
        
        let projectionMatrix = matrix_perspective_left_hand(PI / 4.0, Float(size.width) / Float(size.height), 0.1, 100.0)
        uniform.projectionMatrix = projectionMatrix
    }
    
    func draw(in view: MTKView) {
        let commandBuffer = commandQueue.makeCommandBuffer()!
        
        let renderPassDescriptor = MTLRenderPassDescriptor()
        renderPassDescriptor.colorAttachments[0].texture = view.currentDrawable!.texture
        renderPassDescriptor.colorAttachments[0].loadAction = .clear
        renderPassDescriptor.colorAttachments[0].storeAction = .store
        renderPassDescriptor.colorAttachments[0].clearColor = MTLClearColorMake(0.2, 0.3, 0.3, 1.0)
        
        renderPassDescriptor.depthAttachment.texture = depthTexture
        renderPassDescriptor.depthAttachment.loadAction = .clear
        renderPassDescriptor.depthAttachment.storeAction = .store
        renderPassDescriptor.depthAttachment.clearDepth = 1.0
    
        let renderEncoder = commandBuffer.makeRenderCommandEncoder(descriptor: renderPassDescriptor)!
        renderEncoder.label = "cube render"
        
        renderEncoder.setViewport(MTLViewport(originX: 0, originY: 0,
                                              width: Double(viewportSize.x),
                                              height: Double(viewportSize.y),
                                              znear: 0.0, zfar: 1.0))
        
        renderEncoder.setRenderPipelineState(renderPipelineState)
        renderEncoder.setDepthStencilState(lessDepthState)

        for i in 0..<cubeMesh.vertexBuffers.count {
            let mtkMeshBuffer = cubeMesh.vertexBuffers[i]
            renderEncoder.setVertexBuffer(mtkMeshBuffer.buffer,
                                          offset: mtkMeshBuffer.offset,
                                          index: Int(VertexInputIndexPosition.rawValue))
        }
        
        renderEncoder.setVertexBytes(&uniform,
                                     length: MemoryLayout<Uniforms>.stride,
                                     index: Int(VertexInputIndexUniforms.rawValue))
        try! renderEncoder.setFragmentTexture(containerTexture)

        // draw cube
        for i in 0..<cubeMesh.submeshes.count {
            let submesh = cubeMesh.submeshes[i]
            renderEncoder.drawIndexedPrimitives(type: submesh.primitiveType,
                                                indexCount: submesh.indexCount,
                                                indexType: submesh.indexType,
                                                indexBuffer: submesh.indexBuffer.buffer,
                                                indexBufferOffset: submesh.indexBuffer.offset)
        }
        
        
        // draw sky
        renderEncoder.setRenderPipelineState(skyBoxPipelineState)
        renderEncoder.setDepthStencilState(lessEqualDepthState)
        
        renderEncoder.setVertexBuffer(skyBoxBuffer)
        
        renderEncoder.setVertexBytes(&uniform,
                                     length: MemoryLayout<Uniforms>.stride,
                                     index: Int(VertexInputIndexUniforms.rawValue))
        renderEncoder.setFragmentTexture(skyBoxTexture, index: Int(CubeMapFragmentInputIndexCubeMap.rawValue))
        
        renderEncoder.drawPrimitives(type: .triangle, vertexStart: 0, vertexCount: skyBoxBuffer.count)
        
        renderEncoder.endEncoding()
        
        commandBuffer.present(view.currentDrawable!)
        commandBuffer.commit()
        commandBuffer.waitUntilCompleted()
        
    }
    
    
}
